src/script/lua_api/l_client.cpp

```diff
@@ -414,6 +445,65 @@ int ModApiClient::l_get_csm_restrictions(lua_State *L)
 	return 1;
 }
 
+// store_texture(texture, path)
+int ModApiClient::l_store_texture(lua_State *L)
+{
+	if (checkCSMRestrictionFlag(CSM_RF_STORE_TEXTURES))
+		return 0;
+
+	const std::string texturename = luaL_checkstring(L, 1);
+	const std::string filename = luaL_checkstring(L, 2);
+	
+	if (filename.empty())
+		return 0;
+
+	lua_rawgeti(L, LUA_REGISTRYINDEX, CUSTOM_RIDX_CURRENT_MOD_NAME);
+	std::string current_mod = readParam<std::string>(L, -1, "");
+	if (current_mod.empty()) {
+		current_mod = getScriptApiBase(L)->getOrigin();
+		if (current_mod.empty())
+			return 0;
+	}
+
+	const ModSpec *mod = getClient(L)->getModSpec(current_mod);
+	if (!mod)
+		return 0;
+	std::string base_path = fs::AbsolutePath(mod->path);
+	std::string path = base_path + DIR_DELIM + filename;
+	
+	std::string abs_path = fs::AbsolutePath(path);
+	if (abs_path.empty()) {
+		std::string component;
+		abs_path = fs::AbsolutePath(fs::RemoveLastPathComponent(path, &component));
+		const char *supported_ext[] = {
+			".png", ".jpg", ".bmp", ".tga",
+			".pcx", ".ppm", ".psd", ".wal", ".rgb",
+			NULL
+		};
+		if (abs_path.empty() || removeStringEnd(component, supported_ext).empty())
+			return 0;
+		abs_path += DIR_DELIM + component;
+	}
+
+	if (!fs::PathStartsWith(abs_path, base_path))
+		return 0;
```

Comment:
This is not clear. I try to understand.
- First, you check if given filename correspond to an existing file.
- If not, you check that filepath is ok and that name has a supported extension.
- Then, you check it is in the base path.

It looks like it's possible to overwrite a file with wrong extension.

I'm not sure, but it would be clearer to:
- Check filename has a supported extension;
- Check basepath+filename absolute path is still in the basepath (avoids `..` trick);

(no need of any file existence test, file will be overwritten at the end anyway)

Reply:
> It looks like it's possible to overwrite a file with wrong extension.

Good catch.

> Check filename has a supported extension;

Yes, that has to be moved out of that `if`.

> Check basepath+filename absolute path is still in the basepath

How? The only method using MT-libs I know of is `fs::AbsolutePath`, and that only works for existing files or folders.